### PR TITLE
Feat: Add Go-sdk Actor Support

### PR DESCRIPTION
**Pull request by LaurenceLiZhixin**
I will finish this pr after the following procedures:
1. Actor invoke with raw data success ✅
2. Timer, Reminders, start and stop actor 's  API support ✅
3. client and  server sides user friendly apis，serialization support ✅
4. actor state support ✅
fix basic invoke support of #21

### CODE REVIEW

**Comment by LaurenceLiZhixin** [comment]
Is adding action feature for go sdk a meaningful work? Does my current implementation generally meet your expectations? @yaron2

**Comment by yaron2** [comment]
> Is adding action feature for go sdk a meaningful work? Does my current implementation generally meet your expectations? @yaron2

That's very meaningful! please take a look at the issue for my design comments. overall looking good.

**Comment by yaron2** [comment]
also FYI tests are failing (build)

**Comment by LaurenceLiZhixin** [comment]
@yaron2 I updated my pr and code review is needed, Thanks.

**Comment by yaron2** [comment]
@LaurenceLiZhixin tests are failing.

**Comment by LaurenceLiZhixin** [comment]
@yaron2 I updated and make ci tests passed.

**Comment by yuridevx** [comment]
Hi, all. We're also waiting for this PR, any plans for when this will be released ? And if any help is needed, such as improving test coverage or something? @yaron2  @LaurenceLiZhixin

**Comment by LaurenceLiZhixin** [comment]
> how are you progressing with state?

I'm working on this part. Just a moment, please

**Comment by LaurenceLiZhixin** [comment]
@yaron2 I added actor state api.

File: actor/client/client.go
Comment (LaurenceLiZhixin) [inline actor/client/client.go:10]: I adding this is to ask, should we split "actor client" from grpcClient interface? or just use the old one, adding these actor api to it?

Comment (yaron2) [reply actor/client/client.go:10]: Let's add the actors API to the existing client.

Comment (yaron2) [reply actor/client/client.go:10]: ping @LaurenceLiZhixin

Comment (LaurenceLiZhixin) [reply actor/client/client.go:10]: > Let's add the actors API to the existing client.

OK

**Comment by LaurenceLiZhixin** [comment]
I updated comments of interface, and samples. Remove unused client interface.

File: actor/actor.go
Comment (daixiang0) [inline actor/actor.go:48]: `b.stateManager = stateManger`

File: actor/actor.go
Comment (daixiang0) [inline actor/actor.go:84]: No need to consider error?

File: actor/codec/codec.go
Comment (daixiang0) [inline actor/codec/codec.go:3]: Remove alias please.

File: actor/config/config.go
Comment (daixiang0) [inline actor/config/config.go:25]: `for _, opt := range opts {`

File: actor/codec/impl/json.go
Comment (daixiang0) [inline actor/codec/impl/json.go:1]: Not sure yaml is as common as json in actor field, if so, please add it.

File: actor/error/error.go
Comment (daixiang0) [inline actor/error/error.go:3]: Need expose?

File: actor/manager/container.go
Comment (daixiang0) [inline actor/manager/container.go:28]: Please handle error.

File: actor/manager/container.go
Comment (daixiang0) [inline actor/manager/container.go:63]: Always success?

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:68]: `returnValue, err := m.getAndCreateActorContainerIfNotExist(actorID).Invoke(methodName, request)`

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:83]: Could we handle rather than panic?

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:128]: change like `log.Printf("failed to unmarshal reminder param, err: %v ", err)`, same else.

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:131]: `_, err :=`

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:136]:
```
s:= &Service{}
```

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:141]: use `fmt`

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:142]: Could we handle error here? Same else.

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:154]: `reflctValue` and `reflectType` are more meaningful.

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:171]: Is it designed to skip error?

File: actor/manager/manager.go
Comment (daixiang0) [inline actor/manager/manager.go:195]: Please do not use magic number.

File: actor/state/state_async_provider.go
Comment (daixiang0) [inline actor/state/state_async_provider.go:79]: Format as `// TODO(@author) comment`, same else.

File: client/client.go
Comment (daixiang0) [inline client/client.go:111]: `RegisterActorTimer(ctx context.Context, req *RegisterActorTimerRequest) error`, same below.

File: examples/actor/README.md
Comment (daixiang0) [inline examples/actor/README.md:22]: Remove this line, same below.

File: examples/actor/README.md
Comment (daixiang0) [inline examples/actor/README.md:70]: Remove extra empty lines, same else.

File: examples/actor/client/main.go
Comment (daixiang0) [inline examples/actor/client/main.go:92]: Could we use a timer to replace sleep?

**Review by daixiang0 — CHANGES_REQUESTED** [review]
Thank you for great contributions! There are some nits and error cases need to handle, if we can not handle error now, please add TODO comment.

File: actor/manager/manager.go
Comment (LaurenceLiZhixin) [reply actor/manager/manager.go:68]: aerr is not  error type, it is ActorErr. 
If change to `err`, it would confict with err below.

File: actor/manager/manager.go
Comment (LaurenceLiZhixin) [reply actor/manager/manager.go:171]: Yes, for illegal methods, we do not export.
I thought I could add a log for illegal method to mention here.

**Comment by LaurenceLiZhixin** [comment]
Thank you very much for your suggestions 🙏. I have revised them one by one. 
For `aerr`, it distinguishes between actor-defined err types and normal errors.

File: actor/codec/impl/yaml.go
Comment (daixiang0) [inline actor/codec/impl/yaml.go:19]: `func (y *YamlCodec) Marshal(v interface{}) ([]byte, error) {`

**Review by daixiang0 — COMMENTED** [review]
@yaron2 please do the second review. lgtm.

**Comment by yaron2** [comment]
@LaurenceLiZhixin the `validate-examples` check is failing. After that's fixed I believe we're good to merge!

File: actor/codec/impl/yaml.go
Comment (daixiang0) [reply actor/codec/impl/yaml.go:19]: Please fix.

**Comment by yaron2** [comment]
@LaurenceLiZhixin the test is still failing.

@daixiang0 can you maybe assist here?

**Comment by daixiang0** [comment]
@LaurenceLiZhixin seems still fail, please fix :)

**Comment by LaurenceLiZhixin** [comment]
@daixiang0 Please approve the workflow

**Comment by daixiang0** [comment]
Please try to replace `timer` with `sleep`, as `sleep` easily cause flaky tests.

**Comment by LaurenceLiZhixin** [comment]
I finally found the reason for the problem that had been bothering me for such a long time.
During the stop procedure of the Validataion test, dapr stop --app-id kills a process that is not listening on port 8080, so that port is not released when the next Example starts, causing the next Example to fail.
I fixed the problem in the README of each example. Please restart the workflow and confirm.
@yaron2  @daixiang0

File: actor/codec/codec.go
Comment (LaurenceLiZhixin) [reply actor/codec/codec.go]: Thanks, fixed.

**Comment by daixiang0** [comment]
@LaurenceLiZhixin That sounds like a bug for `dapr stop`, your fix is a workaround indeed.

**Comment by LaurenceLiZhixin** [comment]
The latest problem is understandable. I didn't give the server enough time to pull dependencies and start, so I extended the start time to fix. @daixiang0

**Review by daixiang0 — APPROVED** [review]
LGTM.

/cc @yaron2

**Review by yaron2 — APPROVED** [review]

**Comment by yaron2** [comment]
Merged, thanks @LaurenceLiZhixin this is an awesome PR.

**Pull request merged**
